src/compiler/parser.ts

```diff
@@ -3791,7 +3791,18 @@ namespace ts {
                     }
                 }
                 else {
-                    leftOperand = makeBinaryExpression(leftOperand, parseTokenNode(), parseBinaryExpressionOrHigher(newPrecedence));
+                    const op = parseTokenNode<Token<BinaryOperator>>();
+                    if (isParenthesizedArrowFunctionExpression() !== Tristate.False) {
+                        const maybeArrow = lookAhead(tryParseParenthesizedArrowFunctionExpression);
+                        if (maybeArrow) {
+                            parseErrorAtRange(maybeArrow, Diagnostics.Invalid_arrow_function_arguments_parentheses_around_the_arrow_function_may_help);
+                        }
+                    }
+                    else if (lookAhead(isStartOfSimpleArrowFunction)) {
+                        const maybeArrow = tryParseAsyncSimpleArrowFunctionExpression() || lookAhead(() => parseSimpleArrowFunctionExpression(<Identifier>parseBinaryExpressionOrHigher(/*precedence*/ 0)));
+                        parseErrorAtRange(maybeArrow, Diagnostics.Invalid_arrow_function_arguments_parentheses_around_the_arrow_function_may_help);
+                    }
+                    leftOperand = makeBinaryExpression(leftOperand, op, parseBinaryExpressionOrHigher(newPrecedence));

```

Comment:
You're currently ignoring it if it's present, but I think that you actually want to gracefully parse the arrow functions you've been able to find above.